### ISSUE TITLE
[feature request, idea] Fused torch.sincos(x) or cossin(x) (maybe alias to complex exponential)

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#90559]
### 🚀 The feature, motivation and pitch

Discussed here with motivation: https://discuss.pytorch.org/t/compute-sin-and-cos-simultaneously-on-the-same-input/167896 (thanks @ptrblck for a very interesting nvfuser response!: https://discuss.pytorch.org/t/compute-sin-and-cos-simultaneously-on-the-same-input/167896/2?u=vadimkantorov)

Some concerns:
1) Can be different designs, probably related to typicaly memory layouts and typical usecase, e.g:
```python
s, c = torch.sincos(x) # returns tuple, need to torch.cat for typical positional encoding case
# or
s, c = torch.sincos(x).unbind(0) # all below return tensor, can be unbound
# or
s, c = torch.sincos(x).unbind(-1)
# or
s, c = torch.sincos(x, dim = k).unbind(k)
```
2)  variants (1) and (2) can be implemented by complex exponential `c, s = torch.exp(1j*x).view_as_real().unbind(-1)`, but it's unclear if `1j*x` will incur a reallocation just for constructing the complex tensor, right?. Does it make sense to add to tensor various complex-related bits: imag, real (?), phase (?), mag (?) - similar to neg or transposed bit - so that a complex-accepting functions can have a specialized path with fewer memory accesses

cc @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh

**Comment by lezcano** [comment]
`1j*x` would indeed create a larger tensor with a bunch of zeros... but this would be fused in PyTorch 2.0, so I would not worry too much in that case :)

**Comment by vadimkantorov** [comment]
cossin(x) / sincos(x) might be a nice discoverable alias then following https://man7.org/linux/man-pages/man3/sincos.3.html :) 

but will it pytorch 2.0 elide the copy? and avoid computing exp(0.0) (as magnitude)?

**Comment by lezcano** [comment]
There's just one way to know!

Running
```python
# sincos.py
import torch
from torch._dynamo import optimize


@optimize("inductor")
def sincos(x):
    return torch.cat((x.sin(), x.cos()), dim=-1)


x = torch.randn(64, 64, device="cuda")
sincos(x)
```
with `TORCHINDUCTOR_TRACE=1 python sincos.py`, generates a `torchdynamo_debug/` folder. In the `output_code.py` file within that folder we can see that it spits the following triton code:


<details>
<summary>
Generated triton code
</summary>

```python

from ctypes import c_void_p, c_long
import torch
import random
from torch import empty_strided, as_strided, device
from torch._inductor.codecache import AsyncCompile

aten = torch.ops.aten
assert_size_stride = torch._C._dynamo.guards.assert_size_stride
async_compile = AsyncCompile()

import triton
import triton.language as tl
from torch._inductor.triton_ops.autotune import grid
from torch._C import _cuda_getCurrentRawStream as get_cuda_stream


triton_fused_cos_sin_0 = async_compile.triton('''
import triton
import triton.language as tl
from torch._inductor.ir import ReductionHint
from torch._inductor.ir import TileHint
from torch._inductor.triton_ops.autotune import pointwise
from torch._inductor.utils import instance_descriptor

@pointwise(size_hints=[4096], filename=__file__, meta={'signature': {0: '*fp32', 1: '*fp32', 2: '*fp32', 3: 'i32'}, 'device': 0, 'constants': {}, 'mutated_arg_names': set(), 'configs': [instance_descriptor(divisible_by_16=(0, 1, 2, 3), equal_to_1=())]})
@triton.jit
def triton_(in_ptr0, out_ptr0, out_ptr1, xnumel, XBLOCK : tl.constexpr):
    xnumel = 4096
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.reshape(tl.arange(0, XBLOCK), [XBLOCK])
    xmask = xindex < xnumel
    x2 = xindex
    x0 = xindex % 64
    x1 = (xindex // 64)
    tmp0 = tl.load(in_ptr0 + (x2), xmask)
    tmp1 = tl.sin(tmp0)
    tmp2 = tl.cos(tmp0)
    tl.store(out_ptr0 + (x0 + (128*x1) + tl.zeros([XBLOCK], tl.int32)), tmp1, xmask)
    tl.store(out_ptr1 + (x0 + (128*x1) + tl.zeros([XBLOCK], tl.int32)), tmp2, xmask)
''')


async_compile.wait(globals())
del async_compile

def call(args):
    arg0_1, = args
    args.clear()
    buf2 = empty_strided((64, 128), (128, 1), device='cuda', dtype=torch.float32)
    buf0 = as_strided(buf2, (64, 64), (128, 1))  # alias
    buf1 = as_strided(buf2, (64, 64), (128, 1), 64)  # alias
    stream0 = get_cuda_stream(0)
    triton_fused_cos_sin_0.run(arg0_1, buf0, buf1, 4096, grid=grid(4096), stream=stream0)
    del arg0_1
    return (buf2, )


if __name__ == "__main__":
    from torch._dynamo.testing import rand_strided
    from torch._inductor.utils import print_performance
    arg0_1 = rand_strided((64, 64), (64, 1), device='cuda', dtype=torch.float32)
    print_performance(lambda: call([arg0_1]))
```
</details>

The generated function is called `call`, and it's clear that, yes, it fuses all the operation into one kernel that copies `sin(x)` and `cos(x)` in-place into (views of) a given buffer that is then returned.

**Comment by vadimkantorov** [comment]
wow, this looks like magic :) will it be able to backprop gradients through this compiled function as well / generate gradient program and compile it too?

**Comment by vadimkantorov** [comment]
For the sake of readability of these triton printouts, it might be good to have more type hints in that generated code: e.g. for `in_ptr0`, `out_ptr0`, `out_ptr1` and some intermediate variables

**Comment by lezcano** [comment]
You can play around with the template above to figure out whether that or other ops are optimised in certain ways or not :)

**Comment by jansel** [comment]
> wow, this looks like magic :) will it be able to backprop gradients through this compiled function as well / generate gradient program and compile it too?

Yes, backwards/gradients will work.

> For enhanced readability of these triton printouts, it might be good to have more type hints in that generated code: e.g. for in_ptr0, out_ptr0, out_ptr1 and some intermediate variables

There are some configs that might make things more readable:

https://github.com/pytorch/pytorch/blob/60e196c241e95f4699855f3b2c34d5a0da8faefc/torch/_inductor/config.py#LL63

https://github.com/pytorch/pytorch/blob/60e196c241e95f4699855f3b2c34d5a0da8faefc/torch/_inductor/config.py#L170

There are disabled by default because they cause compile cache misses which increases compile times.

**Comment by ngimel** [comment]
That said, the correct way for torchinductor (and for nvfuser) to generate this op would be to use sincos intrinsic, to avoid repeated exp(0) computations https://docs.nvidia.com/cuda/libdevice-users-guide/__nv_sincosf.html#__nv_sincosf, but we'd need to indroduce sincos op and lowering to make that happen. @vadimkantorov is that something you would be interested in? Python only, no c++.

**Comment by vadimkantorov** [comment]
@ngimel Would that be similar to work in https://github.com/pytorch/pytorch/pull/92632?

**Comment by ngimel** [comment]
Yes, but it looks like libdevice doesn't have sincos

**Comment by ngimel** [comment]
Closing, `sincos` can be easily implemented with torch.compile

**Comment by vadimkantorov** [comment]
> Yes, but it looks like libdevice doesn't have sincos

nvidia's libdeivce does have it, no? https://docs.nvidia.com/cuda/libdevice-users-guide/__nv_fast_sincosf.html#__nv_fast_sincosf

you're talking about triton's libdevice bindings, right? https://github.com/openai/triton/blob/main/python/triton/language/libdevice.py indeed does not seem to have it yet. so the way forward would be to first have it libdevice.py, right?

**Comment by ngimel** [comment]
Do you see any real perf difference when using `sincos` or `sin` and `cos` separately? This is a bandwidth bound operation, unless you fuse a lot of other stuff on top of it, it doesn't matter whether there's a fast `sincos` or separate `sin` and `cos`.
